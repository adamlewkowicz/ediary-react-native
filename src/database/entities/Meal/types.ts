import { DateDay } from '../../../types';

export type GetMacroHistoryResult = {
  day: DateDay
  carbs: number
  prots: number
  fats: number
  kcal: number
}

export type GetMacroSummaryResult = {
  data: GetMacroHistoryResult[]
  average: {
    carbs: number
    prots: number
    fats: number
    kcal: number
  }
}