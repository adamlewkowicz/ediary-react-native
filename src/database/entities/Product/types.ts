import { ProductId } from '../../../types';
import { Product } from '.';

export interface FindMostProductIdsResult {
  count: number
  productId: ProductId
}

export interface FindMostUsedResult extends FindMostProductIdsResult {
  product: Product
}