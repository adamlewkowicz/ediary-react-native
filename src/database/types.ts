import { Omit } from 'utility-types';

export type EntityType<E> = Omit<E, 'hasId' | 'save' | 'remove' | 'reload'>;