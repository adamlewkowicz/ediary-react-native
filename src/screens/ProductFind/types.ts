import { Product } from '../../database/entities';
import { NormalizedProduct } from '../../services/IlewazyApi/types';

export type FoundProductType = 'saved' | 'fetched';

export type FoundProducts = (
  {
    type: 'saved'
    data: Product
  } | 
  {
    type: 'fetched'
    data: NormalizedProduct
  }
)[];