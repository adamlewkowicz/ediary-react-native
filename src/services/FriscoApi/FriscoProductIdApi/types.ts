
export interface Response {
  productId: FriscoProductId
  seoData: {
    title: string
    description: string
    imgAlt: string
  },
  description: string
  officialProductName: string
  brandbank: Brandbank[]
}

type Brandbank = NutritionBrandbank | ProducerBrandbank | IngredientsBrandbank

interface BrandbankLayout {
  sectionId: number
  sectionName: string
  fields: {
    fieldId: number
    fieldName: string
  }[]
}

export interface NutritionBrandbank extends BrandbankLayout {
  sectionId: 2
  sectionName: 'Wartości odżywcze'
  fields: MacroField[]
}

export interface MacroField {
  fieldId: 85
  fieldName: 'Wartości odżywcze'
  content: {
    Headings: string[]
    Nutrients: {
      Name: NutritionName
      Values: string[]
    }[]
  }
}

interface ProducerBrandbank extends BrandbankLayout {
  sectionId: 3
  sectionName: 'Informacje producenta'
  fields: [
    {
      fieldId: 25
      fieldName: 'Marka'
      content: string[]
    }
  ]
}

interface IngredientsBrandbank extends BrandbankLayout {
  sectionId: 1
  sectionName: 'Składniki'
  fields: ({
    fieldId: 178
    fieldName: 'Spis składników odżywczych na przodzie opakowania'
    contentType: 'FrontOfPackGDA'
    content: {
      Reference: 'Na porcję'
      Headers: string[]
      'Wartość energetyczna': {
        /** 2,3 kJ */
        Quantity: string
        /** '0.00' */
        Percentage: string
        Rating: null | string
      }
      Footers: string[]
    }[]
  } | {
    fieldId: 84
    fieldName: 'Składniki'
    contentType: 'LongTextItems'
    content: string[]
  })[]
}

type NutritionName =
  | 'Energia'
  | 'Tłuszcz '
  | ' w tym kwasy nasycone '
  | 'Węglowodany '
  | ' w tym cukry '
  | 'Białko '
  
export interface FriscoProductId extends String {}