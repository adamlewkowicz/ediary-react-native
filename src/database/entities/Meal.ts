import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  OneToMany,
  ManyToOne,
  JoinColumn,
  Between,
} from 'typeorm';
import { Product, IProduct, IProductMerged } from './Product';
import { MealProduct } from './MealProduct';
import { MealId, UserId, DateDay, ProductUnit, ProductId } from '../../types';
import { User } from './User';
import { USER_ID_UNSYNCED } from '../../common/consts';
import { GenericEntity } from './Generic';
import { DeepPartial } from 'typeorm';
import { EntityType } from '../types';

@Entity('Meal', {
  name: 'meals'
})
export class Meal extends GenericEntity {

  @PrimaryGeneratedColumn()
  id!: MealId;

  @Column()
  name!: string;

  @Column('decimal', { precision: 5, scale: 2, default: 0 })
  carbs!: number

  @Column('decimal', { precision: 5, scale: 2, default: 0 })
  prots!: number
  
  @Column('decimal', { precision: 5, scale: 2, default: 0 })
  fats!: number

  @Column('decimal', { precision: 5, scale: 2, default: 0 })
  kcal!: number

  @Column('text', { default: () => 'CURRENT_TIMESTAMP' })
  date!: string;

  @OneToMany(type => MealProduct, mealProduct => mealProduct.meal, {
    onDelete: 'CASCADE'
  })
  mealProducts!: MealProduct[]

  @Column('number', { default: USER_ID_UNSYNCED })
  userId!: UserId | null;

  @ManyToOne(type => User)
  @JoinColumn({ name: 'userId' })
  user!: User;

  static findByDay(dateDay: DateDay) {
    return this.find({
      where: { date: Between(`${dateDay} 00:00:00`, `${dateDay} 23:59:59`) },
      relations: ['mealProducts', 'mealProducts.product']
    });
  }

  static async addAndCreateProduct(
    mealId: MealId,
    payload: DeepPartial<Product>,
    quantity: number = 100
  ): Promise<IProduct> {
    const product = await Product.save(payload);
    const mealProduct = await MealProduct.save({
      productId: product.id,
      unit: product.unit,
      quantity,
      mealId
    });

    return { ...product, ...mealProduct };
  }

  static async addProduct(
    mealId: MealId,
    productId: ProductId,
    quantity: number = 100,
    unit: ProductUnit = 'g'
  ): Promise<{ product: IProductMerged, created: boolean }> {
    const [product, mealProduct] = await Promise.all([
      Product.findOneOrFail(productId),
      MealProduct.findOne({ mealId, productId })
    ]);
    const [portion] = product.portions;
    const portionValue = portion ? portion.value : quantity;

    if (mealProduct) {
      mealProduct.quantity += portionValue;
      await mealProduct.save();
      const mergedProduct = { ...mealProduct, ...product };
      return { product: mergedProduct, created: false  };
    } else {
      const createdMealProduct = await MealProduct.save({
        quantity: portionValue,
        mealId,
        productId,
        unit
      });
      const mergedProduct = { ...createdMealProduct, ...product };
      return { product: mergedProduct, created: true };
    }
  }

}

export type IMeal = EntityType<Meal>;