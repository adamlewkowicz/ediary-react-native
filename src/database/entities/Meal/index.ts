import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  OneToMany,
  ManyToOne,
  JoinColumn,
  Between,
} from 'typeorm';
import { Product, IProductMerged } from '../Product';
import { MealProduct } from '../MealProduct';
import { MealId, UserId, DateDay, ProductId } from '../../../types';
import { User } from '../User';
import { DeepPartial } from 'typeorm';
import { EntityType } from '../../types';
import { GenericEntity } from '../../generics/GenericEntity';
import dayjs from 'dayjs';
import { DATE_FORMAT, DATE_DAY } from '../../../common/consts';
import { Macro } from '../../embeds/Macro';
import { DiaryTemplate } from '../../../store/reducers/diary';
import { getDayFromDate } from '../../../common/utils';
import { GetMacroSummaryResult, GetMacroHistoryResult } from './types';
import { baseMacro } from '../../../common/helpers';

@Entity('meal')
export class Meal extends GenericEntity {

  @PrimaryGeneratedColumn()
  id!: MealId;

  @Column()
  name!: string;

  @Column(type => Macro)
  macro!: Macro;

  @Column('text', { default: () => 'CURRENT_TIMESTAMP' })
  date!: string;

  @OneToMany(
    type => MealProduct,
    mealProduct => mealProduct.meal,
    { cascade: true }
  )
  mealProducts?: MealProduct[];

  @Column('number', { default: null, nullable: true })
  userId!: UserId | null;

  @ManyToOne(type => User, { nullable: true })
  @JoinColumn({ name: 'userId' })
  user!: User;

  static createWithDate(
    payload: DeepPartial<IMeal>,
    date: Date
  ): Promise<Meal> {
    const formattedDate = dayjs(date).format(DATE_FORMAT);
    return this.save({
      ...payload,
      date: formattedDate
    });
  }

  static findByDay(dateDay: DateDay) {
    return this.find({
      where: { date: Between(`${dateDay} 00:00:00`, `${dateDay} 23:59:59`) },
      relations: ['mealProducts', 'mealProducts.product']
    });
  }

  static async addAndCreateProduct(
    mealId: MealId,
    payload: DeepPartial<Product>,
    quantity: number = 100
  ): Promise<IProductMerged> {
    const product = await Product.save(payload);
    const mealProduct = await MealProduct.save({
      productId: product.id,
      quantity,
      mealId
    });
    return { ...product, ...mealProduct };
  }

  static async addProduct(
    mealId: MealId,
    productId: ProductId,
    quantity?: number
  ): Promise<{ product: IProductMerged, rawProduct: Product, action: 'create' | 'update' }> {
    const product = await Product.findOneOrFail(productId);
    const mealProduct = await MealProduct.findOne({ mealId, productId });
    const portionValue = quantity ? quantity : product.portion;
    const rawProduct = product;

    if (mealProduct) {
      mealProduct.quantity += portionValue;
      await mealProduct.save();
      const mergedProduct = { ...mealProduct, ...product };
      return { product: mergedProduct, rawProduct, action: 'update' };
    } else {
      const createdMealProduct = await MealProduct.save({
        quantity: portionValue,
        mealId,
        productId
      });
      const mergedProduct = { ...createdMealProduct, ...product };
      return { product: mergedProduct, rawProduct, action: 'create' };
    }
  }

  static async createWithProduct(
    payload: DeepPartial<Meal>,
    productId: ProductId,
    quantity?: number
  ): Promise<Meal> {
    const product = await Product.findOneOrFail(productId);

    const createdMeal = await Meal.save(payload);
    await MealProduct.save({
      mealId: createdMeal.id,
      productId: product.id,
      quantity: quantity || product.portion
    });
    
    const mealWithRelations = await Meal.findOneOrFail(createdMeal.id, {
      relations: ['mealProducts', 'mealProducts.product']
    });
    return mealWithRelations; 
  }

  static createFromTemplate(
    template: DiaryTemplate,
    date: Date,
    productId: ProductId,
    quantity?: number,
  ): Promise<Meal> {
    const newMeal = {
      name: template.name,
      date: dayjs(date).format(`${DATE_DAY} ${template.time}`)
    }
    return Meal.createWithProduct(
      newMeal,
      productId,
      quantity
    );
  }

  static async getMacroHistory(
    startDay: DateDay,
    periodInDays = 7,
  ): Promise<GetMacroHistoryResult[]> {
    const datePeriod = dayjs(startDay as any).add(periodInDays, 'day');
    const endDay = getDayFromDate(datePeriod);

    const result: GetMacroHistoryResult[] = await Meal
      .createQueryBuilder('meal')
      .select('DATE(meal.date)', 'day')
      .addSelect('SUM(meal.carbs)', 'carbs')
      .addSelect('SUM(meal.prots)', 'prots')
      .addSelect('SUM(meal.fats)', 'fats')
      .addSelect('SUM(meal.kcal)', 'kcal')
      .where('meal.date >= :startDate', { startDate: `${startDay} 00:00:00` })
      .andWhere('meal.date <= :endDate', { endDate: `${endDay} 23:59:00` })
      .groupBy('day')
      .getRawMany();

    return result;
  }

  static async getMacroSummary(
    startDay: DateDay,
    periodInDays?: number,
  ): Promise<GetMacroSummaryResult> {
    const macroHistory = await this.getMacroHistory(startDay, periodInDays);

    const macroAverages = macroHistory.reduce((macro, record, index, self) => {
      const summedMacro = {
        carbs: macro.carbs + record.carbs,
        prots: macro.prots + record.prots,
        fats: macro.fats + record.fats,
        kcal: macro.kcal + record.kcal
      }
      if (index === self.length - 1) {
        return {
          carbs: summedMacro.carbs / self.length,
          prots: summedMacro.prots / self.length,
          fats: summedMacro.fats / self.length,
          kcal: summedMacro.kcal / self.length 
        }
      }
      return summedMacro;
    }, { ...baseMacro });

    const macroSummary: GetMacroSummaryResult = {
      data: macroHistory,
      average: macroAverages
    }

    return macroSummary;
  }

}

export type IMeal = EntityType<Meal>;
export type IMealRequired = IMeal