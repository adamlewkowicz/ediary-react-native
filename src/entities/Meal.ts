import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  OneToMany,
  BaseEntity,
  ManyToOne,
  JoinColumn,
} from 'typeorm/browser';
import { Product } from './Product';
import { productRepository, mealProductRepository } from '../repositories';
import { MealProduct } from './MealProduct';
import { MealId, UserId } from '../types';
import { User } from './User';
import { USER_ID_UNSYNCED } from '../common/consts';

@Entity('Meal', {
  name: 'meals'
})
export class Meal extends BaseEntity {

  @PrimaryGeneratedColumn()
  id!: MealId;

  @Column()
  name!: string;

  @Column('decimal', { precision: 5, scale: 2, default: 0 })
  carbs!: number

  @Column('decimal', { precision: 5, scale: 2, default: 0 })
  prots!: number
  
  @Column('decimal', { precision: 5, scale: 2, default: 0 })
  fats!: number

  @Column('decimal', { precision: 5, scale: 2, default: 0 })
  kcal!: number

  @Column('text', { default: () => 'CURRENT_TIMESTAMP' })
  date!: string;

  @OneToMany(type => MealProduct, mealProduct => mealProduct.meal, {
    onDelete: 'CASCADE'
  })
  mealProducts!: MealProduct[]

  @Column('number', { default: USER_ID_UNSYNCED })
  userId!: UserId | null;

  @ManyToOne(type => User)
  @JoinColumn({ name: 'userId' })
  user!: User;

  async addProduct(
    productId: Product['id']
  ) {
    const mealId = this.id;
    const product = await productRepository().findOne(productId);

    if (!product) {
      throw new Error(
        `Product with id ${productId} doesn't exist`
      );
    }

    const mealProduct = await mealProductRepository().save({
      productId: product.id,
      mealId
    });

    return { ...product, ...mealProduct };
  }

  async addAndCreateProduct(
    payload: Product
  ) {
    const mealId = this.id;
    const product = await productRepository().save(payload);
    const quantity = 100;
    const unit = 'g';

    const mealProduct = await mealProductRepository().save({
      productId: product.id,
      mealId,
      quantity,
      unit
    });

    return { ...product, ...mealProduct };
  }

}