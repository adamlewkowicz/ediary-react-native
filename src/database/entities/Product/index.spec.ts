import { Product } from '.'
import { Meal } from '../Meal';
import { NormalizedProduct } from '../../../services/IlewazyApi/types';

describe('Product', () => {

  describe('findMostUsed()' ,() => {
    it('should find most used products', async () => {
      const product = await Product.save({ name: 'Cucumber' });
      await Meal.createWithProductId({ name: 'Cucumber soup' }, product.id);

      const [foundProduct] = await Product.findMostUsed();
    
      expect(foundProduct.productId).toBe(product.id);
      expect(foundProduct.count).toBe(1);
    });
  });

  describe('findRecentlyUsed()', () => {
    it('should find recently used products', async () => {
      const product = await Product.save({ name: 'Cucumber' });
      await Meal.createWithProductId({ name: 'Cucumber soup' }, product.id);

      const [foundProduct] = await Product.findRecentlyUsed();
    
      expect(foundProduct.id).toBe(product.id);
    });
  });

  describe('saveNormalizedProduct()', () => {
    it('should properly parse and save normalized product', async () => {
      const normalizedProduct: NormalizedProduct = {
        _id: 1,
        name: 'Abc',
        carbs: 0,
        prots: 0,
        fats: 0,
        kcal: 0,
        portions: [],
      }
      const saveSpy = jest.spyOn(Product, 'save');
      const result = await Product.saveNormalizedProduct(normalizedProduct);

      expect(saveSpy).toHaveBeenCalledTimes(1);
      expect(result).toMatchSnapshot({
        id: expect.any(Number),
        updatedAt: expect.any(String),
        createdAt: expect.any(String),
      });
    });
  });
  
});