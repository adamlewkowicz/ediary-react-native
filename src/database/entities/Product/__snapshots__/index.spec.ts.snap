// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product saveNormalizedProduct() should properly parse and save normalized product 1`] = `
Object {
  "barcode": null,
  "createdAt": Any<String>,
  "id": Any<Number>,
  "images": Array [],
  "isVerified": true,
  "macro": Macro {
    "carbs": 0,
    "fats": 0,
    "kcal": 0,
    "prots": 0,
  },
  "name": "Abc",
  "portions": Array [],
  "unit": "g",
  "updatedAt": Any<String>,
  "userId": null,
}
`;
